src/babel-serializer/geometry/obj.cc

```diff
@@ -1,2 +1,417 @@
+
+namespace
+{
+template <class ScalarT>
+babel::obj::geometry<ScalarT> read_impl(cc::span<const std::byte> data, babel::obj::read_config const& cfg, babel::error_handler on_error)
+{
+    using namespace babel;
+
+    obj::geometry<ScalarT> geometry;
+
+    cc::vector<cc::string_view> active_groups = {"default"};
+    int group_point_start = 0;
+    int group_line_start = 0;
+    int group_face_start = 0;
+
+    size_t prev_pos = 0;
+    size_t pos = 0;
+    auto const next_line = [&]() -> cc::string_view {
+        prev_pos = pos;
+        if (pos >= data.size())
+        {
+            on_error(data, data.last(0), "expected line", severity::error);
+        }
+
+        // skip comment lines
+        while (pos < data.size() && char(data[pos]) == '#')
+        {
+            while (pos < data.size() && char(data[pos]) != '\n')
+                ++pos;
+            if (pos < data.size() && char(data[pos]) == '\n')
+                ++pos;
+        }
+
+        auto p_start = reinterpret_cast<char const*>(data.data() + pos);
+        while (pos < data.size())
+        {
+            auto c = char(data[pos]);
+            if (c == '\n' || c == '#')
+                break;
+
+            ++pos;
+        }
+        auto p_end = reinterpret_cast<char const*>(data.data() + pos);
+
+        // ignore rest of comment
+        while (pos < data.size())
+        {
+            auto c = char(data[pos]);
+            if (c == '\n')
+                break;
+
+            ++pos;
+        }
+
+        // point behind newline
+        if (pos < data.size() && char(data[pos]) == '\n')
+            ++pos;
+
+        // trim whitespaces
+        while (pos < data.size() && cc::is_space(char(data[pos])))
+            ++pos;
+
+        return cc::string_view(p_start, p_end);
+    };
+
+    auto const has_more_lines = [&]() -> bool { return pos < data.size(); };
+
+    auto const parse_int = [&](cc::string_view s) -> int {
+        cc::string ss = s; // null-terminated
+        int i;
+        auto const r = std::sscanf(ss.c_str(), "%d", &i);
+        if (r == 0)
+        {
+            on_error(data, cc::as_byte_span(s), "unable to parse int", severity::error);
+            return 0;
+        }
+        return i;
+    };
+
+    auto const parse_float = [&](cc::string_view s) -> ScalarT {
+        cc::string ss = s; // null-terminated
+        ScalarT f;
+        if constexpr (std::is_same_v<ScalarT, float>)
+        {
+            auto const r = std::sscanf(ss.c_str(), "%f", &f);
+            if (r == 0)
+            {
+                on_error(data, cc::as_byte_span(s), "unable to parse float", severity::error);
+                return 0;
+            }
+        }
+        else
+        {
+            auto const r = std::sscanf(ss.c_str(), "%lf", &f);
+            if (r == 0)
+            {
+                on_error(data, cc::as_byte_span(s), "unable to parse double", severity::error);
+                return 0;
+            }
+        }
+        return f;
+    };
```

Comment:
take a look at `<clean-core/from_string>` (it's implementation is not ideal but at least the API is proper)

Reply:
Ahh that's better. This was essentially copied from the pcd parser.

Reply:
Eh yeah that parser is older than from_string :D I should update it